src/main/ts/rules/TypesAtTop.ts

```diff
@@ -0,0 +1,276 @@
+import { ESLintUtils, TSESTree, TSESLint } from '@typescript-eslint/utils';
+
+const createRule = ESLintUtils.RuleCreator(
+  () => 'https://github.com/tinymce/eslint-plugin'
+);
+
+type MessageIds = 'typeNotAtTop';
+type Options = [];
+
+interface FileAnalysis {
+  readonly lastImportLine: number;
+  readonly typeDeclarations: readonly TSESTree.Node[];
+  readonly codeStatements: readonly TSESTree.Node[];
+}
+
+const TYPE_DECLARATION_NODE_TYPES: TSESTree.AST_NODE_TYPES[] = [
+  TSESTree.AST_NODE_TYPES.TSInterfaceDeclaration,
+  TSESTree.AST_NODE_TYPES.TSTypeAliasDeclaration,
+  TSESTree.AST_NODE_TYPES.TSEnumDeclaration
+];
+
+const CODE_STATEMENT_NODE_TYPES: TSESTree.AST_NODE_TYPES[] = [
+  TSESTree.AST_NODE_TYPES.VariableDeclaration,
+  TSESTree.AST_NODE_TYPES.FunctionDeclaration,
+  TSESTree.AST_NODE_TYPES.ClassDeclaration,
+  TSESTree.AST_NODE_TYPES.ExpressionStatement,
+  TSESTree.AST_NODE_TYPES.ExportDefaultDeclaration
+];
+
+/**
+ * Checks if a node is a type declaration (interface, type alias, or enum).
+ * Handles both standalone and exported type declarations.
+ */
+const isTypeDeclaration = (node: TSESTree.Node): boolean => {
+  if (TYPE_DECLARATION_NODE_TYPES.includes(node.type)) {
+    return true;
```

Comment:
As the AI points out on your tinymce PR, TypeScript is not this simple. We do some weird things in our codebases.